src/Bounds.cpp

```diff
@@ -576,24 +581,35 @@ class Bounds : public IRVisitor {
         } else if (can_prove(b.min == b.max)) {
             Expr e1 = a.has_lower_bound() ? a.min / b.min : a.min;
             Expr e2 = a.has_upper_bound() ? a.max / b.max : a.max;
-            // TODO: handle real numbers with can_prove(b.min > 0) and can_prove(b.min < 0) as well - treating floating point as
-            // reals can be error prone when dealing with division near 0, so for now we only consider integers in the can_prove() path
-            if (op->type.is_uint() || is_positive_const(b.min) || (op->type.is_int() && can_prove(b.min >= 0))) {
-                interval = Interval(e1, e2);
-            } else if (is_negative_const(b.min) || (op->type.is_int() && can_prove(b.min <= 0))) {
-                if (e1.same_as(Interval::neg_inf())) {
-                    e1 = Interval::pos_inf();
-                }
-                if (e2.same_as(Interval::pos_inf())) {
-                    e2 = Interval::neg_inf();
-                }
-                interval = Interval(e2, e1);
-            } else if (a.is_bounded()) {
-                // Sign of b is unknown.
-                Expr cmp = b.min > make_zero(b.min.type().element_of());
-                interval = Interval(select(cmp, e1, e2), select(cmp, e2, e1));
-            } else {
+
+            Type t = op->type.element_of();
+
+            if (op->type.is_int() && op->type.bits() <= 16 &&
+                (!a.has_lower_bound() || !can_prove(a.min != t.min())) &&
```

Comment:
I would find this much easier to understand if the bounds_of_type case was the else case of this if. !can_prove(!badthing) -> general case is just harder to wrap your head around than can_prove(!badthing) -> special case